job-runner/chain/traverser.go

```diff
@@ -383,100 +440,74 @@ func (t *traverser) runJobs() {
 				t.runnerRepo.Remove(job.Id)
 			}()
 
-			// Retrieve job and sequence try info from the chain for the Runner.
-			sequenceTries := t.chain.SequenceTries(job.Id) // used in job logs
-			totalJobTries := t.chain.TotalTries(job.Id)    // used in job logs
-			// When resuming a stopped job, only try the job
-			// [allowed tries - tries before being stopped] times, so the total
-			// number of times the job is tried (during this sequence try) stays
-			// correct. The job's last try (the try it was stopped on) doesn't
-			// count, so subtract 1 if it was tried at least once before
-			// being stopped.
-			triesBeforeStopped := uint(0)
-			if job.State == proto.STATE_STOPPED {
-				triesBeforeStopped = t.chain.LatestRunTries(job.Id)
-				if triesBeforeStopped > 0 {
-					triesBeforeStopped--
-				}
+			// Increment sequence try count if this is sequence start job, which
+			// currently means sequenceId == job.Id.
+			if t.chain.IsSequenceStartJob(job.Id) {
+				t.chain.IncrementSequenceTries(job.Id, 1)
+				jLogger.Infof("sequence try %d", t.chain.SequenceTries(job.Id))
 			}
 
-			runner, err := t.rf.Make(job, t.chain.RequestId(), totalJobTries, triesBeforeStopped, sequenceTries)
+			// Job tries for current sequence try and total tries for all seq tries.
+			// For new chains, these are zero. For suspended/resumed chains they can
+			// be > 0 which is why we pass them to the job runner: to resume for the
+			// last counts.
+			curTries, totalTries := t.chain.JobTries(job.Id)
+
+			runner, err := t.rf.Make(job, t.chain.RequestId(), curTries, totalTries)
 			if err != nil {
 				// Problem creating the job runner - treat job as failed.
 				// Send a JobLog to the RM so that it knows this job failed.
+				atomic.AddInt64(&t.pending, -1)
 				job.State = proto.STATE_FAIL
 				err = fmt.Errorf("problem creating job runner: %s", err)
 				t.sendJL(job, err)
```

Comment:
This isn't a bug from you, but if creating the runner fails on a resumed or retried job (i.e. total job tries is already > 0, and we've already sent some logs for this job), then we're sending a duplicate job log (one with this Try # has already been sent).
eg: job was run once and failed, so total job tries = 1 and the RM already has a log with `Try: 1`. We retry the sequence and this time around we get an error creating runner, so we send another job log with `Try: 1`.
A simple fix seems to be incrementing job tries by 1 here.

Reply:
Yeah, or maybe not sending a JLE? Technically, a JLE = the job ran (successful or not), but in this can't we can't really say it ran because we didn't get that far (can't run without a runner).

Reply:
Yeah, but if we don't send a JLE it's very confusing if someone looks back at the logs and sees "job A completed but then job B was never run?!". You could end up with a request that's Failed, but no job logs that show Failed.